src/System.Management.Automation/engine/Subsystem/SubsystemInfo.cs

```diff
@@ -0,0 +1,322 @@
+        {
+            return new SubsystemInfoImpl<TConcreteSubsystem>(kind) {
+                AllowUnregistration = allowUnregistration,
+                AllowMultipleRegistration = allowMultipleRegistration,
+            };
+        }
+
+        internal static SubsystemInfo Create<TConcreteSubsystem>(
+            SubsystemKind kind,
+            bool allowUnregistration,
+            bool allowMultipleRegistration,
+            ReadOnlyCollection<string> requiredCmdlets,
+            ReadOnlyCollection<string> requiredFunctions) where TConcreteSubsystem : class, ISubsystem
+        {
+            return new SubsystemInfoImpl<TConcreteSubsystem>(kind) {
+                AllowUnregistration = allowUnregistration,
+                AllowMultipleRegistration = allowMultipleRegistration,
+                RequiredCmdlets = requiredCmdlets,
+                RequiredFunctions = requiredFunctions,
+            };
+        }
+
+        #endregion
+
+        #region "ImplementationInfo"
+
+        /// <summary>
+        /// Information about an implementation of a subsystem.
+        /// </summary>
+        public class ImplementationInfo
+        {
+            internal ImplementationInfo(ISubsystem implementation)
+            {
+                Id = implementation.Id;
+                Kind = implementation.Kind;
+                Name = implementation.Name;
+                Description = implementation.Description;
+                ImplementationType = implementation.GetType();
+            }
+
+            /// <summary>
+            /// Gets the unique identifier for a subsystem implementation.
+            /// </summary>
+            public Guid Id { get; }
+
+            /// <summary>
+            /// Gets the kind of subsystem.
+            /// </summary>
+            public SubsystemKind Kind { get; }
+
+            /// <summary>
+            /// Gets the name of a subsystem implementation.
+            /// </summary>
+            public string Name { get; }
+
+            /// <summary>
+            /// Gets the description of a subsystem implementation.
+            /// </summary>
+            public string Description { get; }
+
+            /// <summary>
+            /// Gets the implementation type.
+            /// </summary>
+            public Type ImplementationType { get; }
+        }
+
+        #endregion
+    }
+
+    internal sealed class SubsystemInfoImpl<TConcreteSubsystem> : SubsystemInfo
+        where TConcreteSubsystem : class, ISubsystem
+    {
+        private ReadOnlyCollection<TConcreteSubsystem> _registeredImpls;
+
+        internal SubsystemInfoImpl(SubsystemKind kind)
+            : base(kind, typeof(TConcreteSubsystem))
+        {
+            _registeredImpls = Utils.EmptyReadOnlyCollection<TConcreteSubsystem>();
+        }
+
+        private protected override void AddImplementation(ISubsystem rawImpl)
+        {
+            lock (_syncObj)
```

Comment:
> We should document these readonly-ish semantics

Agreed. I will add more comments.

> ImmutableArray 

I thought about it at the time, but chose to not use it to avoid depending on `System.Collections.Immutable.dll` at startup. When working on the experimental flag feature, I initially used `ImmutableDictionary` but then changed to a custom type `ReadOnlyBag` defined in our code working on improving the startup time, to avoid loading of `System.Collections.Immutable` at the startup.

In the subsystem scenario, registration operations will be minimum, and in most cases, the registered implementation will never be unregistered, so optimization for reading is important but not for mutating operations.

Reply:
> It sounds like a custom ReadOnlyBag-like collection to encapsulate the logic (and possibly optimise allocations) would be a good way to go

Yeah, if we find that we need to further memory-optimize this code.